server/mailet/integration-testing/src/test/java/org/apache/james/smtp/extensions/SMTPHeloHooksTest.java

```diff
@@ -0,0 +1,182 @@
+import org.apache.james.modules.protocols.ImapGuiceProbe;
+import org.apache.james.modules.protocols.SmtpGuiceProbe;
+import org.apache.james.probe.DataProbe;
+import org.apache.james.smtp.extensions.hooks.DeclinedHeloHook;
+import org.apache.james.smtp.extensions.hooks.DenyHeloHook;
+import org.apache.james.smtp.extensions.hooks.DenySoftHeloHook;
+import org.apache.james.smtp.extensions.hooks.OkHeloHook;
+import org.apache.james.smtp.extensions.hooks.RecordingHeloHook;
+import org.apache.james.utils.DataProbeImpl;
+import org.apache.james.utils.IMAPMessageReader;
+import org.apache.james.utils.SMTPMessageSender;
+import org.apache.james.utils.SMTPSendingException;
+import org.junit.After;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+public class SMTPHeloHooksTest {
+    private static final String FROM = "fromuser@" + DEFAULT_DOMAIN;
+    private static final String TO = "to@" + DEFAULT_DOMAIN;
+
+    @Rule
+    public SMTPMessageSender messageSender = new SMTPMessageSender(DEFAULT_DOMAIN);
+    @Rule
+    public TemporaryFolder temporaryFolder = new TemporaryFolder();
+    @Rule
+    public StaticInputChecker resultChecker = new StaticInputChecker();
+    @Rule
+    public IMAPMessageReader imapMessageReader = new IMAPMessageReader();
+
+    private TemporaryJamesServer jamesServer;
+
+    private void createJamesServer(SmtpConfiguration.Builder smtpConfiguration) throws Exception {
+        MailetContainer.Builder mailetContainer = TemporaryJamesServer.SIMPLE_MAILET_CONTAINER_CONFIGURATION
+            .putProcessor(CommonProcessors.deliverOnlyTransport());
+
+        jamesServer = TemporaryJamesServer.builder()
+            .withBase(MemoryJamesServerMain.SMTP_AND_IMAP_MODULE)
+            .withSmtpConfiguration(smtpConfiguration)
+            .withMailetContainer(mailetContainer)
+            .build(temporaryFolder.newFolder());
+
+        DataProbe dataProbe = jamesServer.getProbe(DataProbeImpl.class);
+        dataProbe.addDomain(DEFAULT_DOMAIN);
+        dataProbe.addUser(FROM, PASSWORD);
+        dataProbe.addUser(TO, PASSWORD);
+    }
+
+    @After
+    public void tearDown() {
+        if (jamesServer != null) {
+            jamesServer.shutdown();
+        }
+    }
+
+    @Test
+    public void heloHookShouldBeCalledWithTheRightArgument() throws Exception {
+        createJamesServer(SmtpConfiguration.builder()
+            .addHook(RecordingHeloHook.class.getCanonicalName()));
+
+        messageSender.connect(LOCALHOST_IP, jamesServer.getProbe(SmtpGuiceProbe.class).getSmtpPort())
+            .sendMessage(FROM, TO);
+
+        awaitAtMostOneMinute.until(() -> resultChecker.getResults().size() > 0);
+        assertThat(resultChecker.getResults())
+            .containsExactly(Pair.of(RecordingHeloHook.class, DEFAULT_DOMAIN));
+    }
+
+    @Test
+    public void mailShouldBeWellDeliveredUponDeclinedHeloHook() throws Exception {
```

Comment:
why? (maybe I don't understand what is a DeclinedHeloHook...)

Reply:
Decline => skipMe Hook

I know that is counter intuitive but you know, I cannot change everything at once on my spare time ;-)